### PR TITLE
feat(observability): Add rate limited debug messages

### LINKED ISSUE #628: Rate limited logging audit
Now that #494 is merged, we can implement rate limited logs throughout Vector. This is _very_ helpful for logging in the hot path and providing valuable insight for users. @LucioFranco performed extensive benchmarking on this and it appears to have a negligible impact on performance. As such, the recommendations here are based on that information.

**tracing**:

- [x] Update to new released versions of tracing https://github.com/timberio/vector/issues/929

**docs**

- [ ] ~~Write docs on how to trouble shoot sinks, sources, and transforms using tracing.~~ (#970)

**sources**

- [x] `file` source - line exceeds `max_line_bytes` - once every 30 seconds
- [x] `stdin` source - line exceeds `max_line_bytes` - once every 30 seconds
- [x] `syslog` source - parsing failed - https://github.com/timberio/vector/pull/885#issuecomment-532385824 - once every 10 seconds
- [x] `tcp` source - line exceeds `max_line_bytes` - once every 30 seconds

**transforms**

- [x] `grok_parser` transform - parsing missed - once every 30 seconds
- [x] `json_parser` transform - parsing failed - once every 30 seconds
- [x] `log_to_metric` transform - `increment_by_value` parsing failed - once every 30 seconds
- [x] `lua` transform - any script / compile error - once every 30 seconds
- [x] `regex_parser` transform - parsing missed - once every 30 seconds
- [x] `regex_parser` transform - field coercion failed - once every 30 seconds
- [x] `tokenizer` transform - field coercion failed - once every 30 seconds

**sinks**

- [x] `aws_cloudwatch_logs` sink - event interpolation (partition keys) missing event field - once every 30 seconds
- [x] `aws_cloudwatch_logs` sink - event interpolation (partition keys) bad type (can't coerce to string) - once every 30 seconds
- [x] `aws_kinesis` sink - event interpolation (partition keys) missing event field - once every 30 seconds
- [x] `aws_s3` sink - event interpolation (partition keys) missing event field - once every 30 seconds
- [x] `elasticsearch` sink - event interpolation (partition keys) missing event field - once every 30 seconds
- [x] `http` sink - event interpolation (partition keys) missing event field - once every 30 seconds

**buffers**

This section is a little tricky, because I want to continually notify users if there is continuous back pressure.

- [x] Log when a buffer becomes full - once every 10 seconds.
- [x] Log when an event is dropped (shed) - once every 10 seconds.

### LINKED ISSUE #806: Update sink encoding to be non optional
~~Most sinks allow the user to specify an encoding type. If this option is not provided the sinks will attempt to check each event and encode it based on if it was explicitly structured by the user. Due to precidence the current implementation of the match statement will attempt to encode the event as a structured event even if the user explicitly included the `encoding = "text"` option. This should be fixed by forcing the left tuple match of the boolean value to be `None`.~~

Update:

Based on the comments below, the sinks will be modified to contain explicit options to encode their events. This means that encoding is no longer optional and a user must select their desired encoding or the sink can not be built.

Implementation wise this should be a simple fix to remove the second tuple item in this match statement https://github.com/timberio/vector/blob/master/src/sinks/aws_cloudwatch_logs/mod.rs#L268 and to change the encoding config item to not be wrapped in an option. In a separate issue, we can consolidate encoding options.

Sinks that need this fix:
- [x] aws_cloudwatch_logs
- [x] aws_s3
- [x] aws_kinesis_streams
- [x] kafka
- [x] splunk
- [x] tcp
- [x] All sinks should have tests for this behavior

### LINKED ISSUE #929: Update tracing dependencies
We might need to update `tracing` deps in accordance with the recent [changes](https://github.com/tokio-rs/tracing/pull/323).

I've run into this while trying to run `cargo outdated`:
```sh
$ cargo outdated
error: no matching package named `tracing-env-logger` found
location searched: https://github.com/tokio-rs/tracing
required by package `vector v0.4.0-dev (/var/folders/8q/vw_18tzd78l41qtzpv8zmn6m0000gq/T/cargo-outdatedPT9OD8)`
```

### LINKED ISSUE #981: Configuration errors are logged twice
Attempting to start `vector` without a configuration file produces:

```
Oct 07 19:08:22.742  INFO vector: Log level "info" is enabled.
Oct 07 19:08:22.743  INFO vector: Loading config. path="/etc/vector/vector.toml"
Oct 07 19:08:22.743 ERROR vector: Config file not found in path. path=/etc/vector/vector.toml     
Oct 07 19:08:22.743 ERROR vector: Config file not found in path. path=/etc/vector/vector.toml
```

This is also true when the configuration file is empty:

```
Oct 07 19:09:02.301  INFO vector: Log level "info" is enabled.
Oct 07 19:09:02.301  INFO vector: Loading config. path="/etc/vector/vector.toml"
Oct 07 19:09:02.303 ERROR vector: Configuration error: missing field `sources`     
Oct 07 19:09:02.303 ERROR vector: Configuration error: missing field `sources`
```

### PR DESCRIPTION
**Pull request by LucioFranco**
This PR contains three major code changes:

- Update all of our tracing dependencies to the latest versions.
- Update `tracing-limit` to use `tracing_subscriber`'s `Layer` trait.
- Instrument the main code base with rate limited tracing events.

This change introduces the ability for tracing events to be rate limited. The primary reason for including this is for events that may be important for a user to see but may arrive in large bursts. This can cause issues around saturating IO and in general making it difficult to diagnose the issue.

The solution is to allow tracing events to provide a `rate_limit_secs` attribute that specifies a window at which we will only see one log and the rest will be counted. This enables users to know that this event is being recorded but will not flood their view into vector.

```rust
 INFO basic: hello, world! count=0 rate_limit_secs=5
TRACE basic: this field is not rate limited!
 INFO basic: "hello, world!" is being rate limited. rate_limit_secs=5
TRACE basic: this field is not rate limited!
TRACE basic: this field is not rate limited!
TRACE basic: this field is not rate limited!
TRACE basic: this field is not rate limited!
 INFO basic: 5 "hello, world!" events were rate limited. rate_limit_secs=5
```

Closes #628 
Closes #929 
Closes #981
Related to #970

Related PRs #778, https://github.com/tokio-rs/tracing/pull/350, https://github.com/tokio-rs/tracing/pull/368 (this one is not going into this change but related to #970).

Signed-off-by: Lucio Franco <luciofranco14@gmail.com>

### CODE REVIEW

**Comment by binarylogic** [comment]
@LucioFranco doesn't this close https://github.com/timberio/vector/issues/628?

**Comment by binarylogic** [comment]
Also, does this cover https://github.com/timberio/vector/issues/929?

**Comment by lukesteensen** [comment]
Lots going on here.

> ```rust
> INFO basic: hello, world! count=0 rate_limit_secs=5
> ```

Do we actually want to print out the rate limiting config with every one of these log lines? Feels like that could be confusing to users.

I'm also not sure what this new "debug" mode that you've added is. Why is this separate from the normal log level and do we really want to introduce a new concept here?

**Comment by LucioFranco** [comment]
@lukesteensen I don't think we can avoid the logging of `rate_limit_secs` currently. Down the line we maybe able to do ignore it via a config option in the fmt subscriber.

The `debug` mode right now just adds the targets to logging like `INFO vector::sources::kafka:` to each logline, which is disabled by default to make them easier to read. I am happy to remove it just found it useful in debugging.

**Comment by lukesteensen** [comment]
> Down the line we maybe able to do ignore it via a config option in the fmt subscriber.

Could you open an issue for this before merging? I do think it'd be a bit cleaner.

> The `debug` mode right now just adds the targets to logging like `INFO vector::sources::kafka:` to each logline, which is disabled by default to make them easier to read. I am happy to remove it just found it useful in debugging.

My issue with this type of thing is that if it's optional, people won't have it enabled when an issue happens. So if it's useful, we should have it on by default. If it's too wordy and not that useful, we can just leave it out.
